lua_obj.h

```diff
@@ -21,14 +22,82 @@ enum class LuaParam : uint8_t {
 };
 
 class duel;
+template<typename T>
+class owned_lua;
 
 class lua_obj {
+	template<typename T>
+	friend class owned_lua;
 public:
 	duel* pduel{ nullptr };
 	int32_t ref_handle{};
+	int32_t weak_ref_handle{};
 	LuaParam lua_type;
 protected:
 	lua_obj(LuaParam _lua_type, duel* _pduel) :pduel(_pduel), lua_type(_lua_type) {}
+private:
+	int num_ref{};
+	void incr_ref();
+	void decr_ref();
+};
+
+template<typename T>
+class owned_lua {
+	T* obj{};
+	owned_lua<T>& assign(T* other) {
+		if(obj) {
+			obj->decr_ref();
+		}
+		if(obj = other; obj) {
```

Comment:
~~move the assignment to outside the if and check explicitly for `nullptr`, like the operator bool below.~~ Actually, thinking more about it, its not the implicit check that bothers me, as its consistently used in the rest of the class, but the assignment within the if, normally this feature is used to declare new intermediate variables to be used *within* the if.